src/Util.cpp

```diff
@@ -1263,6 +1267,21 @@ to_lowercase(string_view string)
 void
 traverse(const std::string& path, const TraverseVisitor& visitor)
 {
+#if __cplusplus >= 201703L
+  if (std::filesystem::is_directory(path)) {
+    for (auto&& p : std::filesystem::directory_iterator(path)) {
```

Comment:
Using `std::filesystem::directory_iterator` is unfortunately not as efficient as `opendir/readdir` (with `_DIRENT_HAVE_D_TYPE` support) since the `std::filesystem` solution needs to perform a `stat` syscall for each entry to determine its type. Therefore the `std::filesystem` solution needs to be used only if `opendir/readdir` aren't available.

Reply:
I did a quick comparison of opendir vs filesystem as I was surprised to learn that it's slower.
Executed on WSL2 with Ubuntu 18 - no idea what kind of effect that has.
Listing all files in llvm:
* opendir ~105ms
* filesystem ~350ms
* filesystem_recursive ~350ms

Although the recusrive option is like 1000 times more readable I can see that it is in fact slower. Will read up on that some more, but that's off topic.
```
    for (auto&& p : std::filesystem::recursive_directory_iterator(path)) {
      std::string entry = p.path().string();
      visitor(entry, false);
    }
```

Reply:
@jrosdahl do you have any background why filesystem is not using `_DIRENT_HAVE_D_TYPE` / `DTTOIF`? Is it just poor implementation?

Reply:
It seems like my initial comment was incorrect; I probably misread the code and didn't take my time to look at the actual filesystem API. On my system (GCC 9.3, ext4), `std::filesystem::(recursive_)directory_iterator` uses the expected and efficient system calls (i.e., not one `stat` per file entry), so now I'm surprised that you see a difference. :slightly_smiling_face:

Anyway, I think the current code is OK.

Reply:
I've just tried with gcc 9.3 on ext4 and I see lots of stat calls :-(
I'll try opening tickets on clang/gcc. Maybe it's WSL related somehow.

Reply:
Found the problem!
1) there is a "bug" in this submission. Will provide suggestion in a second.
2) I measured the speed of recursive_directory_iterator incorrectly. I had a typo and was calling the directory_iterator version while trying to measure recursive_directory_iterator.

Now listing all llvm files (those times do not jitter at all):
opendir ~130ms
directory_iterator as submitted here ~670ms
directory_iterator fixed ~160ms
recursive_directory_iterator ~140ms

P.S. it's not comparable to the numbers above since I checked out a different commit on llvm, have different build directories etc.

Reply:
Nevermind it's already fixed :) :(

The trick was to not use `std::filesystem::is_directory(path)` and `std::filesystem::is_regular_file(path)` as that info is already available within `p`. `p.is_directory()` and `p.is_regular_file()` return without a stat call.